simplified-ui-catalog/src/main/java/org/nypl/simplified/ui/catalog/saml20/CatalogSAML20ViewModel.kt

```diff
@@ -0,0 +1,216 @@
+
+        this.logger.debug("obtained download info")
+        this.downloadInfo.set(
+          DownloadInfo(
+            mimeType = mimeType
+          )
+        )
+
+        val cookies = WebViewUtilities.dumpCookiesAsAccountCookies(
+          CookieManager.getInstance(),
+          this.webViewDataDir
+        )
+
+        val loginState = account.loginState
+
+        if (loginState is AccountLoginState.AccountLoggedIn) {
+          val credentials = loginState.credentials
+
+          if (credentials is AccountAuthenticationCredentials.SAML2_0) {
+            account.setLoginState(
+              loginState.copy(
+                credentials = credentials.copy(
+                  cookies = cookies
+                )
+              )
+            )
+          }
+        }
+
+        val bookWithStatus = this.bookRegistry.bookOrNull(this.bookID)
+
+        if (bookWithStatus != null) {
+          val book = bookWithStatus.book
+
+          try {
+            this.booksController.bookBorrow(
+              accountID = book.account,
+              entry = book.entry,
+            )
+            Unit
+          } catch (e: Throwable) {
+            this.logger.error("failed to start borrow task: ", e)
+            this.bookRegistry.updateIfStatusIsMoreImportant(
+              BookWithStatus(book, BookStatus.FailedLoan(book.id, this.failMinimal(e)))
+            )
+          }
+        }
+
+        this.eventSubject.onNext(
+          CatalogSAML20Event.Succeeded()
+        )
+      }
+    }
+
+    private fun failMinimal(
+      exception: Throwable
+    ): TaskResult.Failure<Unit> {
+      val recorder = TaskRecorder.create()
+      recorder.beginNewStep("Logging in...")
+      recorder.currentStepFailed(
+        message = exception.message ?: exception.javaClass.canonicalName ?: "unknown",
+        errorCode = "borrowLoginFailed",
+        exception = exception
+      )
+      return recorder.finishFailure()
+    }
+  }
+
+  override fun onCleared() {
+    super.onCleared()
+
+    if (this.downloadInfo.get() == null) {
+      this.logger.debug("no download info obtained; cancelling login")
+
+      val bookWithStatus = this.bookRegistry.bookOrNull(this.bookID)
+
+      if (bookWithStatus != null) {
+        val book = bookWithStatus.book
+        this.booksController.bookDownloadCancel(book.account, this.bookID)
+        this.bookRegistry.update(BookWithStatus(book, BookStatus.fromBook(book)))
+      }
+    }
+
+    this.eventSubject.onComplete()
+  }
+
+  val webViewClient: WebViewClient =
+    CatalogSAML20WebClient(
+      eventSubject = this.eventSubject,
+      logger = this.logger,
+      booksController = this.booksController,
+      bookRegistry = this.bookRegistry,
+      bookID = this.bookID,
+      downloadInfo = this.downloadInfo,
+      account = this.account,
+      webViewDataDir = this.webViewDataDir
+    )
+
+  val isWebViewClientReady: Boolean
+    get() = (this.webViewClient as CatalogSAML20WebClient).isReady
+}
```

Comment:
Looks like there's some significant overlap between the catalog and account fragments and view models. Is this some business logic we might consider consolidating in the future?